### DISCUSSION TITLE
EMQX certificates question

### DISCUSSION THREAD

**Discussion by gjalring** [emqx/emqx#7204] [Q&A]
Hi,

Trying to set up secure communications with EMQX using certificates, but I am bit new to this still.
I have successfully managed to connect client to mqtt server using ssl, but I had to provide both server and client with certificates, in effect two way ssl. 

From what I have understood this should need be necessary as long as the client has the rootCA to verify server cert. But no matter what I do i get the error  "Unable to verify first certificate" in the MQTT X client when trying to connect to server using either "CA signed server" or "Self signed" with only rootCA. If I provide client with cert + key it works, or if i turn off the "SSL secure" option.

I have read that the error I get about "Unable to verify first certificate" could be due to missing intermediate cert, but I have checked and it is present in the certificate. Anyone able to tell me what I am doing wrong?

From what I understand the two way SSL is enabled by these two settings, but I have tried to have them both enabled like below or commented out:
```
listener.ssl.external.verify = verify_none
listener.ssl.external.fail_if_no_peer_cert = false
```

This is my listener config (removed setting that was commented out to shorten it for posting):

```
# MQTT/SSL - External SSL Listener for MQTT Protocol

## listener.ssl.$name is the IP address and port that the MQTT/SSL
## listener will bind.

listener.ssl.external = 8883

listener.ssl.external.acceptors = 16

listener.ssl.external.max_connections = 102400

listener.ssl.external.max_conn_rate = 500

listener.ssl.external.active_n = 100

listener.ssl.external.zone = external

listener.ssl.external.handshake_timeout = 15s

## Path to the file containing the user's private PEM-encoded key.
listener.ssl.external.keyfile = etc/certs/My-key.pem

## Path to a file containing the user certificate.
listener.ssl.external.certfile = etc/certs/MyCert.pem

## Path to the file containing PEM-encoded CA certificates. The CA certificates
## are used during server authentication and when building the client certificate chain.
listener.ssl.external.cacertfile = etc/certs/RootCA.pem

listener.ssl.external.verify = verify_none


listener.ssl.external.fail_if_no_peer_cert = false

listener.ssl.external.ciphers = TLS_AES_256_GCM_SHA384,TLS_AES_128_GCM_SHA256,TLS_CHACHA20_POLY1305_SHA256,TLS_AES_128_CCM_SHA256,TLS_AES_128_CCM_8_SHA256,ECDHE-ECDSA-AES256-GCM-SHA384,ECDHE-RSA-AES256-GCM-SHA384,ECDHE-ECDSA-AES256-SHA384,ECDHE-RSA-AES256-SHA384,ECDHE-ECDSA-DES-CBC3-SHA,ECDH-ECDSA-AES256-GCM-SHA384,ECDH-RSA-AES256-GCM-SHA384,ECDH-ECDSA-AES256-SHA384,ECDH-RSA-AES256-SHA384,DHE-DSS-AES256-GCM-SHA384,DHE-DSS-AES256-SHA256,AES256-GCM-SHA384,AES256-SHA256,ECDHE-ECDSA-AES128-GCM-SHA256,ECDHE-RSA-AES128-GCM-SHA256,ECDHE-ECDSA-AES128-SHA256,ECDHE-RSA-AES128-SHA256,ECDH-ECDSA-AES128-GCM-SHA256,ECDH-RSA-AES128-GCM-SHA256,ECDH-ECDSA-AES128-SHA256,ECDH-RSA-AES128-SHA256,DHE-DSS-AES128-GCM-SHA256,DHE-DSS-AES128-SHA256,AES128-GCM-SHA256,AES128-SHA256,ECDHE-ECDSA-AES256-SHA,ECDHE-RSA-AES256-SHA,DHE-DSS-AES256-SHA,ECDH-ECDSA-AES256-SHA,ECDH-RSA-AES256-SHA,AES256-SHA,ECDHE-ECDSA-AES128-SHA,ECDHE-RSA-AES128-SHA,DHE-DSS-AES128-SHA,ECDH-ECDSA-AES128-SHA,ECDH-RSA-AES128-SHA,AES128-SHA

listener.ssl.external.reuseaddr = true
```

MQTTX connects fine if i turn off SSL secure, but thats no good.
![image](https://user-images.githubusercontent.com/59511471/156713601-38a322c8-0124-4724-a44b-937c1b75bfd6.png)

"Unable to verify first certificate" if i just provide rootCA, if provided Client cert and key it connects.
![image](https://user-images.githubusercontent.com/59511471/156713747-58fb923a-e445-441e-821a-7e3678ed9a65.png)

**Comment by zmstone** [comment]
Hi.
We have made two videos to explain TLS in depth.
https://youtu.be/HRqJLi7-9KU
https://youtu.be/W7SedpZzQdo

**Comment by zmstone** [comment]
listener.ssl.external.verify = verify_peer # validate client certificates 
listener.ssl.external.fail_if_no_peer_cert = true # force client to send its certificates
